src/from_str.rs

```diff
@@ -41,7 +52,81 @@ pub fn expand(input: &DeriveInput, trait_name: &'static str) -> Result<TokenStre
                 Ok(#body)
             }
         }
-    })
+    }
+}
+
+fn enum_from(
+    input: &DeriveInput,
+    state: State,
+    trait_name: &'static str,
+) -> TokenStream {
+    let mut variants_caseinsensitive = HashMap::default();
+    for variant_state in state.enabled_variant_data().variant_states {
+        let variant = variant_state.variant.unwrap();
+        if variant.fields != Fields::Unit {
+            panic!("Only enums with no fields can derive({})", trait_name)
+        }
+
+        variants_caseinsensitive
+            .entry(variant.ident.to_string().to_lowercase())
+            .or_insert_with(Vec::new)
+            .push(variant.ident.clone());
+    }
+
+    let input_type = &input.ident;
+    let visibility = &input.vis;
+
+    let err_name = format_ident!("Parse{}Error", input_type);
+    let err_message =
+        format!("invalid {}", input_type.to_string().to_case(Case::Lower));
+
+    let mut cases = vec![];
+
+    // if a case insensitve match is unique match do that
+    // otherwise do a case sensitive match
+    for (ref canonical, ref variants) in variants_caseinsensitive {
+        if variants.len() == 1 {
+            let variant = &variants[0];
+            cases.push(quote! {
+                #canonical => #input_type::#variant,
+            })
+        } else {
+            for variant in variants {
+                let variant_str = variant.to_string();
+                cases.push(quote! {
+                    #canonical if(src == #variant_str) => #input_type::#variant,
+                })
+            }
+        }
+    }
+
+    let trait_path = state.trait_path;
+
+    quote! {
+
+        #[derive(Debug, Clone, PartialEq, Eq)]
+        #visibility struct #err_name;
```

Comment:
I'm not sure I like creating a type out of nothing here, but I did not like the other options better:

- could use `&'static str` but that does not impelement `std::error::Error` which make it annoying for combining with other errors
- could have a `derive_more::ParseError` but that makes the dependencies complicates as you cant have code and procmarcos in the same dependency